File: src/convex/figure/points/Deq.java
package convex.figure.points;

import convex.geometry.Point;

public class Deq {
    private Point[] array = new Point[16];
    private int size = 0, head = 0, tail = array.length - 1;
    private int forward(int index) {
        return ++index < array.length ? index : 0;
    }
    private int backward(int index) {
        return --index >= 0 ? index : array.length - 1;
    }
    protected int length() {
        return size;
    }
    protected void pushFront(Point p) {
        array[head = backward(head)] = p;
        size += 1;
    }
    protected void pushBack(Point p) {
        array[tail = forward(tail)] = p;
        size += 1;
    }
    protected Point popFront() {
        Point p = front();
        head = forward(head);
        size -= 1;
        return p;
    }
    protected Point popBack() {
        Point p = back();
        tail = backward(tail);
        size -= 1;
        return p;
    }
    protected Point front() {
        return array[head];
    }
    protected Point back() {
        return array[tail];
    }
}

File: Convex/Figure/Points/Deq.cs
namespace Convex.Figure.Points
{
    public class Deq
    {
        public int Length { get; private set; }
        private readonly Geometry.Point[] points;
        private int head;
        private int tail;

        public Deq()
        {
            Length = 0;
            points = new Geometry.Point[16];
            head = 0;
            tail = points.Length - 1;
        }

        public Geometry.Point Front
        {
            get => points[head];
        }

        public Geometry.Point Back
        {
            get => points[tail];
        }

        public void PushFront(Geometry.Point point)
        {
            points[head = Backward(head)] = point;
            Length += 1;
        }

        public void PushBack(Geometry.Point point)
        {
            points[tail = Forward(tail)] = point;
            Length += 1;
        }

        public Geometry.Point PopFront
        {
            get
            {
                var point = Front;
                head = Forward(head);
                Length -= 1;
                return point;
            }
        }

        public Geometry.Point PopBack
        {
            get
            {
                var point = Back;
                tail = Backward(tail);
                Length -= 1;
                return point;
            }
        }

        private int Forward(int index)
        {
            return ++index < points.Length ? index : 0;
        }

        private int Backward(int index)
        {
            return --index >= 0 ? index : points.Length - 1;
        }
    }
}
